source/common/http/async_client_impl.cc

```diff
@@ -79,6 +79,9 @@ void AsyncStreamImpl::encodeHeaders(HeaderMapPtr&& headers, bool end_stream) {
       },
       nullptr);
 #endif
+  if (remote_closed_) {
```

Comment:
IMO we want a guard in data/trailers, but not headers. I don't think there is any reasonable way headers can happen after reset with a substantial bug beyond what we have today.